Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+{
+    public class NuGetCredentialsValidator : ValidatorBase
+    {
+        internal static NuGetCredentialsValidator Instance;
+
+        static NuGetCredentialsValidator()
+        {
+            Instance = new NuGetCredentialsValidator();
+        }
+
+        public static ValidationState IsValidStatic(ref string matchedPattern,
+                                                    ref Dictionary<string, string> groups,
+                                                    ref string failureLevel,
+                                                    ref string message,
+                                                    out Fingerprint fingerprint)
+        {
+            return ValidatorBase.IsValidStatic(Instance,
+                                               ref matchedPattern,
+                                               ref groups,
+                                               ref failureLevel,
+                                               ref message,
+                                               out fingerprint);
+        }
+
+        public static ValidationState IsValidDynamic(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            return IsValidDynamic(Instance,
+                                  ref fingerprint,
+                                  ref message,
+                                  ref options);
+        }
+
+        protected override ValidationState IsValidStaticHelper(ref string matchedPattern,
+                                                               ref Dictionary<string, string> groups,
+                                                               ref string failureLevel,
+                                                               ref string message,
+                                                               out Fingerprint fingerprint)
+        {
+            fingerprint = default;
+
+            if (!groups.TryGetNonEmptyValue("secret", out string secret) ||
+                !groups.TryGetNonEmptyValue("id", out string id) ||
+                !groups.TryGetNonEmptyValue("host", out string host))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            if (StaticStringHelper.LikelyPowershellVariable(secret))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            fingerprint = new Fingerprint
+            {
+                Host = host, // technically, this is XML which could contain multiple hosts.
+                Id = id,
+                Secret = secret,
+                Platform = nameof(AssetPlatform.NuGet),
+            };
+
+            return ValidationState.Unknown;
+        }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            string hostXmlAsString = fingerprint.Host;
+            string username = fingerprint.Id;
```

Comment:
>string username [](start = 12, length = 15)

please extract parameters to variables named id, host, etc. order them from shortest to longest. you can review other rules to observe the consistent style we prefer.

Reply:
I actually wanted to talk to @eddynaka and you about this.  First, I don't understand why we assign these values to variables when we don't modify them and we have random access to the dictionary in the fingerprint.  Was there a reason behind that?

Second, regarding the naming inside the validators; in my view, the fingerprint is deliberately abstract to handle as many options as possible.  But validators know internally what those abstract variables are, and it doesn't need to be so vague.  It makes sense to perform this translation from abstract to concrete if we're going to be creating separate variables for these values.  NuGetCredentialsValidator knows that "id" really means "username" and that "host" is actually a bunch of XML with hosts inside.  This is a good opportunity for self-documenting code, is it not?  We lose a lot of information if we keep the abstract names, don't we?

---
In reply to: [611238606](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611238606) [](ancestors = 611238606)